### DISCUSSION TITLE
Enabling security in CubeFS

### DISCUSSION THREAD

**Discussion by Zorlin** [cubefs/cubefs#3087] [Q&A] [answered]
Hi there,

I want to place some basic restrictions on CubeFS so that machines with access to my local network can't wipe my entire cluster by simply talking to my Master/Resource Manager node.

I followed the directions in [the old ChubaoFS documentation](https://chubaofs.readthedocs.io/en/latest/user-guide/authnode.html) (which should likely be merged into the current CubeFS docs, or at least the portions about authnode) and successfully got authnode up and running including TLS certificates, but despite doing this when I use cfs-cli:
```
wings@blackberry:~$ cat ~/.cfs-cli.json 

{
  "masterAddr": [
    "10.7.1.201:17010"
  ],
  "timeout": 60
}
```

I have access to all CLI options and can delete volumes and do all kinds of stuff with zero authentication.

How do I lock that down? It's the main thing stopping me from rolling out CubeFS properly!

**Comment by Zorlin** [comment]
https://github.com/cubefs/cubefs/pull/1857

This pull request appears to contain the necessary code to make this work :astonished:

It was merged, but the changes don't appear in v3.3.1 yet.

**Reply by leonrayang** [reply]
Thanks for the reminder, will be included in v3.3.2

**Comment by Zorlin** [comment]
I was able to take the branch "release-3.3.1" and cherry pick #1857 onto it, then make a custom CubeFS build.

I'm now at the stage where I have a HA master and HA authnode setup, with TLS and everything working. Clients can only do major operations (or any sort of change really) when authenticated with a valid key now.

However, when I attempt to add datanodes, it claims that the key for "DatanodeService" doesn't exist, even though I created it through the cfs-authtool the same way I created the keys for the master (following the old ChubaoFS docs).

**Comment by Zorlin** [comment]
Logs and details below. The main issue is a **key not exists** error.

```
root@cubefs-authnode01:/etc/cubefs# cat data_datanode.json
{
    "id": "DatanodeService",
    "role": "service",
    "caps": "{\"API\":[\"*:*:*\"]}"
}
````

```
root@cubefs-authnode01:/etc/cubefs# cfs-authtool api -https -host=10.7.1.151:8443 -ticketfile=ticket_admin.json -data=data_datanode.json -output=key_datanode.json AuthService createkey

body: {"code":0,"msg":"success","data":"5<REDACTED>n"}
```

```
root@cubefs-authnode01:/etc/cubefs# cat key_datanode.json
{
  "id": "DatanodeService",
  "auth_key": "<REDACTED=",
  "access_key": "<REDACTED",
  "secret_key": "<REDACTED",
  "create_ts": 1707698243,
  "role": "service",
  "caps": "{\"API\":[\"*:*:*\"]}",
  "auth_id_key": "ey<REDACTED>=="
}
```

```
2024/02/11 23:33:16.752274 [INFO ] http_server.go:230: action[AuthenticationInterceptor] parseAndCheckClientKey failed, RequestURI[/dataNode/add?addr=10.7.1.201:17310&zoneName=default&clientIDKey=e<REDACTED>==], err[get ticket from auth node failed, clientIDKey[e<REDACTED>==], err[request error, code[2], msg[action[GetKey], clusterID[rifflabs-ldn] ID:DatanodeService, err:key not exists ]]]
2024/02/11 23:33:16.752283 [INFO ] api_args_parse.go:1508: URL[/dataNode/add?addr=10.7.1.201:17310&zoneName=default&clientIDKey=e<REDACTED>==],remoteAddr[10.7.1.162:47016],response
```

Config of my datanode.json:
```
{
  "role": "datanode",
  "listen": "17310",
  "prof": "17320",
  "logDir": "/var/log/cubefs/datanode/log",
  "logLevel": "info",
  "raftHeartbeat": "17330",
  "raftReplica": "17340",
  "raftDir":"/var/lib/cubefs/datanode/raft",
  "exporterPort": 9502,
  "masterAddr": [
     "10.7.1.161:17010",
     "10.7.1.162:17010",
     "10.7.1.163:17010"
  ],
  "disks": [
     "/mnt/brick/cubefs:10737418240"
  ],
  "serviceIDKey": "e<REDACTED>=="
}
```

**Reply by leonrayang** [reply]
@M1eyu2018  please take a look at this issue.

**Reply by M1eyu2018** [reply, marked as answer]
@Zorlin hi, the key of datanode may be not stored in authnode. Please check it by using getKey api as follows:

cfs-authtool api -https -host=10.7.1.151:8443 -ticketfile=ticket_admin.json -data=data_datanode.json -output=key_datanode.json AuthService getkey

And if the key is actually not in authnode, just create key by authtool again.

**Reply by Zorlin** [reply]
Hi @M1eyu2018 

In my case I hit a really nasty bug - the key simply did not exist and no amount of deleting and re-creating it would make it work. I could create it, then attempting to create it again would result in a KeyAlreadyExists error (or something like that, I don't recall, sorry), but even though that suggests it now exists - if I poll the API for that key it doesn't exist.

The only way I found to fix it was *wiping my entire authnode cluster* and recreating everything. That did work however and I was able to get CubeFS fully online with authnode enabled :tada:

**Reply by M1eyu2018** [reply]
ok, in your case, if the result of getkey api is also that key not exists, the keystore in authnode may go wrong.

**Reply by Zorlin** [reply]
Yep, I believe that's what happened.

**Comment by Zorlin** [comment]
@leonrayang any thoughts?

**Reply by M1eyu2018** [reply]
@Zorlin hi, the key of datanode may be not stored in authnode. Please check it by using getKey api as follows:

cfs-authtool api -https -host=10.7.1.151:8443 -ticketfile=ticket_admin.json -data=data_datanode.json -output=key_datanode.json AuthService getkey

And if the key is actually not in authnode, just create key by authtool again.

**Comment by leonrayang** [comment]
> Hi there,
> 
> I want to place some basic restrictions on CubeFS so that machines with access to my local network can't wipe my entire cluster by simply talking to my Master/Resource Manager node.
> 
> I followed the directions in [the old ChubaoFS documentation](https://chubaofs.readthedocs.io/en/latest/user-guide/authnode.html) (which should likely be merged into the current CubeFS docs, or at least the portions about authnode) and successfully got authnode up and running including TLS certificates, but despite doing this when I use cfs-cli:
> 
> ```
> wings@blackberry:~$ cat ~/.cfs-cli.json 
> 
> {
>   "masterAddr": [
>     "10.7.1.201:17010"
>   ],
>   "timeout": 60
> }
> ```
> 
> I have access to all CLI options and can delete volumes and do all kinds of stuff with zero authentication.
> 
> How do I lock that down? It's the main thing stopping me from rolling out CubeFS properly!

@Zorlin The authnode documentation will be merged as soon as possible ：）

**Reply by Zorlin** [reply]
Thank you!!

**Comment by major-sam** [comment]
Hi!
How auth implemented in helm deployment?
Is it possible to add docs or comments for it?
Is it possible to implement native k8s secrets for using and storing auth keys?

**Reply by leonrayang** [reply]
> Hi! How auth implemented in helm deployment? Is it possible to add docs or comments for it? Is it possible to implement native k8s secrets for using and storing auth keys?

It is not avaliable now,  Most of the maintainers are not familiar with the components and Helm, we hope that the end users could help to finish it.
